lib/ukboot/boot.c

```diff
@@ -134,43 +149,48 @@ void ukplat_entry(int argc, char *argv[])
 	}
 #endif /* CONFIG_LIBUKLIBPARAM */
 
+
 #if !CONFIG_LIBUKBOOT_NOALLOC
-	/* initialize memory allocator
-	 * FIXME: allocators are hard-coded for now
-	 */
 	uk_pr_info("Initialize memory allocator...\n");
-	ukplat_memregion_foreach(&md, UKPLAT_MEMRF_ALLOCATABLE) {
+#ifdef CONFIG_HAVE_PAGING
+	/* Paging is enabled. The available physical memory has already been
+	 * added to the frame allocator. No heap area has been mapped.
+	 */
+	free_pages  = pt->fa->free_memory >> PAGE_SHIFT;
+	alloc_pages = free_pages - PT_PAGES(free_pages);
+
+	rc = ukplat_page_map(pt, CONFIG_LIBUKBOOT_HEAP_BASE, __PADDR_ANY,
+			     alloc_pages, PAGE_ATTR_PROT_RW, 0);
+	if (unlikely(rc))
+		UK_CRASH("Heap mapping failed: %d\n", rc);
+
+	a = uk_alloc_init((void *)CONFIG_LIBUKBOOT_HEAP_BASE,
+			  alloc_pages << PAGE_SHIFT);
+#else /* CONFIG_HAVE_PAGING */
+	/* Paging is disabled so we still have the static boot page table set
+	 * that maps (some of) the physical memory to virtual memory. The
+	 * FREE memory regions reflect this. Try to use these memory regions to
+	 * initialize the allocator. If it fails, we will try again with the
+	 * next region. As soon we have an allocator, we simply add every
+	 * subsequent region to it.
+	 */
+	ukplat_memregion_foreach(&md, UKPLAT_MEMRT_FREE, 0, 0) {
```

Comment:
Right, any other entry point (e.g. EFI, Linux Boot Protocol, Multiboot v2 etc.) would prepare the memory region list. We can think about merging once in `_ukplat_entry` as you said. Although, if we do it in `ukplat_memregion_list_insert()`, we only have to modify an existing entry instead of removing one (which means a ´memmove`). So probably it is not really faster. Especially, since you basically have to do the same checks.

We add this, when we add further protocols to the mix 👍🏻